tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php

```diff
@@ -322,24 +233,28 @@ protected function createDiscount(string $cartRuleReference, array $data, AddDis
             $command->setCode($data['code']);
         }
 
-        if ($command instanceof AddCartLevelDiscountCommand
-            || $command instanceof AddProductLevelDiscountCommand
-            || $command instanceof AddOrderLevelDiscountCommand
+        if ($command->getDiscountType()->getValue() === DiscountType::CART_DISCOUNT
+            || $command->getDiscountType()->getValue() === DiscountType::PRODUCTS_DISCOUNT
+            || $command->getDiscountType()->getValue() === DiscountType::ORDER_DISCOUNT
         ) {
             if (!empty($data['reduction_percent'])) {
                 $command->setPercentDiscount(new DecimalNumber($data['reduction_percent']));
             }
 
             if (!empty($data['reduction_amount'])) {
-                $command->setAmountDiscount(
-                    new DecimalNumber($data['reduction_amount']),
-                    new CurrencyId($this->getSharedStorage()->get($data['reduction_currency'])),
-                    PrimitiveUtils::castStringBooleanIntoBoolean($data['taxIncluded']),
-                );
+                try {
+                    $command->setAmountDiscount(
+                        new DecimalNumber($data['reduction_amount']),
+                        new CurrencyId($this->getSharedStorage()->get($data['reduction_currency'])),
+                        PrimitiveUtils::castStringBooleanIntoBoolean($data['taxIncluded']),
+                    );
+                } catch (DomainConstraintException $e) {
+                    $this->setLastException(new DiscountConstraintException($e->getMessage(), DiscountConstraintException::INVALID_DISCOUNT_VALUE_CANNOT_BE_NEGATIVE));
+                }
```

Comment:
```suggestion
                } catch (DomainConstraintException $e) {
                    $this->setLastException($e);
                }
```
You're not supposed to translate an exception into another in behat The prupose if error checking is to make sure the domain code throws the expected exception, if the logic is handeld by behat then the domain doesn't do what it's supposed to do It's just behat test validating itself and it has few interest

Either the command code must be adapted so that it throws what you expect it to throw, or the test is not correct It's possible that a previous check made sense before but is not relevant with the refacto as well Then the test logic can evolve according to the code evolution